### PR TITLE
add IconInput

### PR DESCRIPTION
**Pull request by akdetrick**
## Description
Adds `IconInput`

## Screenshots
Screen Shot 2019-12-04 at 11 30 16 AM
Screen Shot 2019-12-04 at 11 30 27 AM

## Checklist
- [x] Docs have been rebuilt (`yarn build:full`)
- [x] All unit tests pass
- [x] Snapshots have been updated
- [x] No lint errors or warnings have been introduced
- [x] **[Version bumped](https://github.com/cbinsights/form-design-system#updating-version-number) if appropriate**

### CODE REVIEW

**Comment by rey-wright** [comment]
My only question here, is why this element wouldn't spit out an input itself, rather than requiring you to wrap an input. 

I sort of get why you did this (composition) but I just wonder.... if this should always be used with an input and nothing else, if there is advantage here this not outputting an input element itself (and using spread props).

File: src/components/style/forms/iconInput.css [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+.fdsIconInput--hasEndIcon input {
+  padding-right: var(--space-double);
+}
+
+.fdsIconInput-icon {
+  position: absolute;
+  top: 1px;
+  bottom: 1px;
+  display: flex;
+  align-items: center;
```
Comment (rey-wright) [inline src/components/style/forms/iconInput.css]: In my own components, I've been using an extra div to utilize `alignChild--center--center`. This is more of a lesson to me, but it seems you're ok curving that class and spelling it out directly here?

File: src/components/forms/iconInput.test.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+      <IconInput
+        IconLeft={StarFilledIcon}
+      >
+        <input type="text" />
+      </IconInput>
+    ).find('.fdsIconInput');
+
+    expect(rootEl.hasClass('fdsIconInput--hasStartIcon')).toBe(true);
+    expect(rootEl.hasClass('fdsIconInput--hasEndIcon')).toBe(false);
+  });
```
Comment (rey-wright) [inline src/components/forms/iconInput.test.js]: I like itttttt

File: src/components/forms/IconInput.jsx [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+IconInput.propTypes = {
+  /** An `input` element child */
+  children: PropTypes.node,
+
+  /**
+   * FDS Icon _reference_
+   * `IconLeft={StarFilledIcon}`
+   */
+  IconLeft: PropTypes.func,
```
Comment (rey-wright) [inline src/components/forms/IconInput.jsx]: 👍 good explanations

File: src/components/forms/IconInput.jsx [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+import FDS from '../../../lib/dictionary/js/styleConstants';
+
+/**
+ * @param {Object} props react props
+ * @returns {ReactElement}
+ */
+const IconInput = ({ IconLeft, IconRight, children }) => {
+  const classNames = cx('fdsIconInput', {
+    'fdsIconInput--hasStartIcon': Boolean(IconLeft),
+    'fdsIconInput--hasEndIcon': Boolean(IconRight),
```
Comment (rey-wright) [inline src/components/forms/IconInput.jsx]: do these need `Boolean` or were you just trying to be more transparent that these values aren't boolean to begin with

File: src/components/style/forms/iconInput.css [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+}
+
+.fdsIconInput-icon--start {
```
Comment (akdetrick) [inline src/components/style/forms/iconInput.css]: i guess these could be utility classes

File: src/components/style/forms/iconInput.css (hunk 1 above)
Comment (akdetrick) [reply src/components/style/forms/iconInput.css]: good call - I'll use the utility

File: src/components/forms/IconInput.jsx (hunk 4 above)
Comment (akdetrick) [reply src/components/forms/IconInput.jsx]: mostly for clarity - it's technically not necessary

**Comment by akdetrick** [comment]
Reasons for using `children` to pass the input:

- Same interface as `DecoratedIcon`
- Avoids putting input-related props on the `IconInput` component (keeps `IconInput` itself smaller and sharper - only takes props related to the icons)
- `<input {...inputProps}` is easier to reason about than `<NotTheInput {...inputProps}` - there's more transparency into which props are going where

**Comment by rey-wright** [comment]
@akdetrick that makes sense... though I'd like to make sure I fully understand, only because if we extrapolate that argument, I'm pretty sure we'd have a bunch more existing FDS components that debatably should be abstracted into multiple components if this is the principle we're abiding by.

**Pull request merged**